cpp/src/arrow/type.cc

```diff
@@ -288,13 +288,25 @@ std::string DurationType::ToString() const {
 // ----------------------------------------------------------------------
 // Union type
 
+constexpr int8_t UnionType::kMaxTypeId;
+constexpr int UnionType::kInvalidChildId;
+
 UnionType::UnionType(const std::vector<std::shared_ptr<Field>>& fields,
-                     const std::vector<uint8_t>& type_codes, UnionMode::type mode)
-    : NestedType(Type::UNION), mode_(mode), type_codes_(type_codes) {
+                     const std::vector<int8_t>& type_codes, UnionMode::type mode)
+    : NestedType(Type::UNION),
+      mode_(mode),
+      type_codes_(type_codes),
+      child_ids_(kMaxTypeId + 1, kInvalidChildId) {
```

Comment:
Do I have it right that this is adding `sizeof(int) * (kMaxTypeId + 1)` (so 512 or 1024 bytes depending on the system) to the footprint of every UnionType instance?

Reply:
Yes. I think this is simpler than having to lazily-initialize it.